### PR TITLE
Enhance the network profiling

### PR DESCRIPTION
**Pull request by mrproliu**
* Optimized eBPF event handler
* Optimize the connection closing event handle process
* Fix traffic direction identification errors

### CODE REVIEW

File: bpf/profiling/network/netmonitor.c [hunk 1]
```diff
@@ -127,63 +153,36 @@ static __always_inline void submit_new_connection(struct pt_regs* ctx, __u32 fun
...
-    // pid is contains
-    if (tgid_should_trace(con->pid) == false) {
-        return;
+    // if the connect event not send, then check the pid or socket family
+    if (con->connect_event_send == false) {
+        // only trace ipv4, v6, or unknown
+        if (family_should_trace(con->socket_family) == false) {
+            return;
+        }
+        // pid is contains
```
Comment (wu-sheng) [inline bpf/profiling/network/netmonitor.c:182]: This comment seems not clear, could you polish it?

File: bpf/profiling/network/netmonitor.c [hunk 2]
```diff
@@ -273,9 +301,14 @@ static __always_inline void process_write_data(void *ctx, __u64 id, struct sock_
         return;
     }
 
-    // pid is contains
-    if (tgid_should_trace(tgid) == false) {
-        return;
+    // if connection not sent
```
Comment (wu-sheng) [inline bpf/profiling/network/netmonitor.c:304]:
```suggestion
    // if connect event is not sent
```

I think you mean this?

File: bpf/profiling/network/netmonitor.c [hunk 3]
```diff
@@ -285,14 +318,20 @@ static __always_inline void process_write_data(void *ctx, __u64 id, struct sock_
...
-            char *buff = _(iovec->iov_base);
-            buf_reader = read_socket_data(buff, bytes_count);
+            struct iovec iov;
+            int err = bpf_probe_read(&iov, sizeof(iov), args->iovec);
+            if (err >= 0) {
+                __u64 size = iov.iov_len;
+                if (size > bytes_count) {
+                    size = bytes_count;
+                }
+//                bpf_printk("iov read buffer size: %lld, connect event send status: %d\n", size, conn->connect_event_send);
```
Comment (wu-sheng) [inline bpf/profiling/network/netmonitor.c:328]: Please remove commented codes if it is pointless. Or log it out if needed.

File: bpf/profiling/network/netmonitor.c [hunk 4]
```diff
@@ -301,10 +340,12 @@ static __always_inline void process_write_data(void *ctx, __u64 id, struct sock_
...
             // if send response data to remote address or receive request data from remote address
             // then, recognized current connection is server
-            } else if ((msg_type == kResponse && data_direction == SOCK_DATA_DIRECTION_INGRESS) ||
-                       (msg_type == kRequest && data_direction == SOCK_DATA_DIRECTION_EGRESS)) {
+            } else if ((msg_type == kResponse && data_direction == SOCK_DATA_DIRECTION_EGRESS) ||
+                       (msg_type == kRequest && data_direction == SOCK_DATA_DIRECTION_INGRESS)) {
                 conn->role = CONNECTION_ROLE_TYPE_SERVER;
             }
+//            bpf_printk("connection: %lld, data_len: %d, original len: %d", conid, buf_reader->data_len, bytes_count);
+//            bpf_printk("message type: %d, data_direction: %d, new role: %d", msg_type, data_direction, conn->role);
```
Comment (wu-sheng) [inline bpf/profiling/network/netmonitor.c:348]: Please remove commented codes if it is pointless. Or log it out if needed.

File: go.mod [hunk 5]
```diff
@@ -8,6 +8,7 @@ require (
 	github.com/hashicorp/go-multierror v1.1.1
 	github.com/hashicorp/golang-lru v0.5.4
 	github.com/ianlancetaylor/demangle v0.0.0-20200824232613-28f6c0f3b639
+	github.com/orcaman/concurrent-map v1.0.0
```
Comment (wu-sheng) [inline go.mod:11]: Is this a new dependency?

Comment (mrproliu) [reply go.mod:11]: Yes, this works to reduce lock contention during map operations in receiving the event from BPF. I will update the license before release.

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**